File: equals.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <iostream>

class Foo {
public:
    Foo() {i=0;}
    Foo(int i) {this->i=i;}
    void setI(int i) {this->i = i;}
    int getI() { return this->i; }
    std::string toString() { return "Foo:" + std::to_string(i); }
    bool operator==(Foo f) { return i == f.i; }
private:
    int i;
};

int main() {

    int i=3;
    if (i==3) { std::cout << "primitive ==" << std::endl; }

    Foo f = Foo(3);
    if (f == Foo(3)) { std::cout << "object == " << std::endl; }

    Foo *fp1 = new Foo(3);
    Foo *fp2 = new Foo(3);
    if (fp1 == fp1) { std::cout << "obvi" << std::endl; }
    if (fp1 == fp2) { std::cout << "nah" << std::endl; }
    if (*fp1 == *fp2) { std::cout << "*nah" << std::endl; }
}

File: Equals.java


public class Equals {
    private int i;

    public Equals(int i) { this.i = i; }
    public void setI(int i) { this.i = i; }
    public String toString() { return "" + i; }
    public boolean equals(Equals o) {return i == o.i;}
    public static void main(String[] args) {

        int i=3;
        if (i == 3) { System.out.println("primitive ==\n"); }

        Equals e = new Equals(3);
        Equals f = new Equals(3);
        if (e == e ) { System.out.println("object e==e\n"); }
        if (e == f ) { System.out.println("object ==\n"); }
        if (e.equals(f)) { System.out.println("object equals()\n"); }

    }

}
